<img src = "images/3_logos.png" width = 600, align = "center">
<br>
<h1 align=center>Tout le monde veut aller au Canada</h1>

Ce notebook explore un jeu de données sur l'immigration au Canada. Les statistiques sont fournies par l'ONU et contiennent des données annuelles sur les flux de migrants enregistrées par les pays de destination.[Source](http://www.un.org/en/development/desa/population/migration/data/empirical2/migrationflows.shtml)

### Pre-process

Importer les librairies et charger le jeu de données dans un dataFrame. Ne charger que la feuille de calcul :"Canada by Citizenship" en supprimant les lignes inutiles en début et en fin de feuille.

In [98]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

canada = pd.read_excel("data/Canada.xlsx",  "Canada by Citizenship", skiprows = 20, nrows = 196)

Afficher les 5 premières lignes.

In [99]:
canada.head()
canada.dropna() 

Type    Coverage  \
0    Immigrants  Foreigners   
1    Immigrants  Foreigners   
2    Immigrants  Foreigners   
3    Immigrants  Foreigners   
4    Immigrants  Foreigners   
5    Immigrants  Foreigners   
6    Immigrants  Foreigners   
7    Immigrants  Foreigners   
8    Immigrants  Foreigners   
9    Immigrants  Foreigners   
10   Immigrants  Foreigners   
11   Immigrants  Foreigners   
12   Immigrants  Foreigners   
13   Immigrants  Foreigners   
14   Immigrants  Foreigners   
15   Immigrants  Foreigners   
16   Immigrants  Foreigners   
17   Immigrants  Foreigners   
18   Immigrants  Foreigners   
19   Immigrants  Foreigners   
20   Immigrants  Foreigners   
21   Immigrants  Foreigners   
22   Immigrants  Foreigners   
23   Immigrants  Foreigners   
24   Immigrants  Foreigners   
25   Immigrants  Foreigners   
26   Immigrants  Foreigners   
27   Immigrants  Foreigners   
28   Immigrants  Foreigners   
29   Immigrants  Foreigners   
..          ...         ...   
166  Immigrants  Foreigners   
167  Immigrants  Foreigners   
168  Immigrants  Foreigners   
169  Immigrants  Foreigners   
170  Immigrants  Foreigners   
171  Immigrants  Foreigners   
172  Immigrants  Foreigners   
173  Immigrants  Foreigners   
174  Immigrants  Foreigners   
175  Immigrants  Foreigners   
176  Immigrants  Foreigners   
177  Immigrants  Foreigners   
178  Immigrants  Foreigners   
179  Immigrants  Foreigners   
180  Immigrants  Foreigners   
181  Immigrants  Foreigners   
182  Immigrants  Foreigners   
183  Immigrants  Foreigners   
184  Immigrants  Foreigners   
185  Immigrants  Foreigners   
186  Immigrants  Foreigners   
187  Immigrants  Foreigners   
188  Immigrants  Foreigners   
189  Immigrants  Foreigners   
190  Immigrants  Foreigners   
191  Immigrants  Foreigners   
192  Immigrants  Foreigners   
193  Immigrants  Foreigners   
194  Immigrants  Foreigners   
195  Immigrants  Foreigners   

                                                OdName  AREA  \
0                                          Afghanistan   935   
1                                              Albania   908   
2                                              Algeria   903   
3                                       American Samoa   909   
4                                              Andorra   908   
5                                               Angola   903   
6                                  Antigua and Barbuda   904   
7                                            Argentina   904   
8                                              Armenia   935   
9                                            Australia   909   
10                                             Austria   908   
11                                          Azerbaijan   935   
12                                             Bahamas   904   
13                                             Bahrain   935   
14                                          Bangladesh   935   
15                                            Barbados   904   
16                                             Belarus   908   
17                                             Belgium   908   
18                                              Belize   904   
19                                               Benin   903   
20                                              Bhutan   935   
21                    Bolivia (Plurinational State of)   904   
22                              Bosnia and Herzegovina   908   
23                                            Botswana   903   
24                                              Brazil   904   
25                                   Brunei Darussalam   935   
26                                            Bulgaria   908   
27                                        Burkina Faso   903   
28                                             Burundi   903   
29                                          Cabo Verde   903   
..                                                 ...   ...   
166                                     

Supprimer les colonnes 'AREA', 'REG', 'DEV', 'Type', 'Coverage' et 'DevName' ne seront pas utiles dans ce notebook.<br>
Renommer ensuite les colonnes 'OdName'=>'Pays' , 'AreaName'=>'Continent' et 'RegName'=>'Region'

In [100]:
canada.drop(['AREA', 'REG', 'DEV', 'Type', 'Coverage', 'DevName'], axis='columns', inplace=True)
print(canada)

                                                OdName  \
0                                          Afghanistan   
1                                              Albania   
2                                              Algeria   
3                                       American Samoa   
4                                              Andorra   
5                                               Angola   
6                                  Antigua and Barbuda   
7                                            Argentina   
8                                              Armenia   
9                                            Australia   
10                                             Austria   
11                                          Azerbaijan   
12                                             Bahamas   
13                                             Bahrain   
14                                          Bangladesh   
15                                            Barbados   
16            

In [101]:
#canada = canada.rename(columns = {'OdName' : 'Pays', 'AreaName' : 'Continent', 'RegName' : 'Region'}, axis=1)

old_names = ['OdName', 'AreaName', 'RegName'] 
new_names = ['Pays', 'Continent', 'Region']
canada.rename(columns=dict(zip(old_names, new_names)), inplace=True)

print(canada)

                                                  Pays  \
0                                          Afghanistan   
1                                              Albania   
2                                              Algeria   
3                                       American Samoa   
4                                              Andorra   
5                                               Angola   
6                                  Antigua and Barbuda   
7                                            Argentina   
8                                              Armenia   
9                                            Australia   
10                                             Austria   
11                                          Azerbaijan   
12                                             Bahamas   
13                                             Bahrain   
14                                          Bangladesh   
15                                            Barbados   
16            

Indexez le dataFrame par la colonne 'Pays>

In [102]:
canada.set_index("Pays")

Continent  \
Pays                                                                                  
Afghanistan                                                                    Asia   
Albania                                                                      Europe   
Algeria                                                                      Africa   
American Samoa                                                              Oceania   
Andorra                                                                      Europe   
Angola                                                                       Africa   
Antigua and Barbuda                                 Latin America and the Caribbean   
Argentina                                           Latin America and the Caribbean   
Armenia                                                                        Asia   
Australia                                                                   Oceania   
Austria                                                                      Europe   
Azerbaijan                                                                     Asia   
Bahamas                                             Latin America and the Caribbean   
Bahrain                                                                        Asia   
Bangladesh                                                                     Asia   
Barbados                                            Latin America and the Caribbean   
Belarus                                                                      Europe   
Belgium                                                                      Europe   
Belize                                              Latin America and the Caribbean   
Benin                                                                        Africa   
Bhutan                                                                         Asia   
Bolivia (Plurinational State of)                    Latin America and the Caribbean   
Bosnia and Herzegovina                                                       Europe   
Botswana                                                                     Africa   
Brazil                                              Latin America and the Caribbean   
Brunei Darussalam                                                              Asia   
Bulgaria                                                                     Europe   
Burkina Faso                                                                 Africa   
Burundi                                                                      Africa   
Cabo Verde                                                                   Africa   
...                                                                             ...   
Swaziland                                                                    Africa   
Sweden                                                                       Europe   
Switzerland                                                                  Europe   
Syrian Arab Republic                                                           Asia   
Tajikistan                                                                     Asia   
Thailand                                                                       Asia   
The former Yugoslav Republic of Macedonia                                    Europe   
Togo                                                                         Africa   
Tonga                                                                       Oceania   
Trinidad and Tobago                                 Latin America and the Caribbean   
Tunisia                                                                      Africa   
Turkey                                                                         Asia   
Turkmenistan                                                                   Asia   
Tuvalu                                                                  

In [103]:
canada['Total'] = canada[canada.columns[2:]].sum(axis=1)
print(canada)

                                                  Pays  \
0                                          Afghanistan   
1                                              Albania   
2                                              Algeria   
3                                       American Samoa   
4                                              Andorra   
5                                               Angola   
6                                  Antigua and Barbuda   
7                                            Argentina   
8                                              Armenia   
9                                            Australia   
10                                             Austria   
11                                          Azerbaijan   
12                                             Bahamas   
13                                             Bahrain   
14                                          Bangladesh   
15                                            Barbados   
16            

Ajoutez une colonne 'Total' affichant le total d'immigrants par pays.

### Visualisations

Paramétrage

In [104]:
# Affichage des graphiques dans le navigateur
%matplotlib inline 

import matplotlib as mpl
import matplotlib.pyplot as plt

mpl.style.use('ggplot') # optionel: pour utiliser le style de ggplot

**1. Line Plot (Courbe)**

Qu'est-ce qu'un *Line Plot* et quand l'utiliser?

Il s'agit ni plus ni moins que d'une série de points de données appelés «marqueurs» reliés par des segments de droite.<br>Ce type de graphique est très adapté aux données continues et c'est aussi le plus adapté pour repésenter les tendances sur une période donnée.

Exemple :

En 2010, Haïti a subi un séisme catastrophique de magnitude 7,0. Le séisme a provoqué des dégâts considérables et des pertes en vies humaines et environ trois millions de personnes ont été touchées par cette catastrophe naturelle. Dans le cadre des efforts humanitaires du Canada, le gouvernement du Canada a intensifié ses efforts pour accepter les réfugiés d'Haïti. Nous pouvons rapidement visualiser cet effort en utilisant un tracé linéaire:

Plusieurs étapes sont nécessaires : 
 - Extraire les données correspondantes à 'Haiti'
 - Transtyper le type associé aux valeurs de l'index en int
 - Créer un graphique de type 'line'
 - Définir le titre : 'Immigration depuis Haiti'
 - Donner un titre aux axes : x : 'Années' et y : 'Nombre d'imigrants'
 - Afficher le graphique

In [105]:
#subset = canada [canada.Pays == 'Haiti']


In [117]:
haiti = canada.iloc[[75]]

#haiti = haiti.stack().reset_index()
haiti.drop(columns=['Pays'], inplace = True)
haiti.drop(haiti.index[[0,1]], inplace = True)
haiti.drop(haiti.tail(1).index,inplace=True)
haiti.columns = ['Years', 'Migrants'] 
haiti['Migrants'] = haiti['Migrants'].astype(int)
haiti.index = haiti.index.map(int)
x=haiti['Years']
y=haiti['Migrants']
plt.plot(x,y)
plt.ylabel("Nombre d'imigrants")
plt.xlabel("Années")
plt.title("Immigration depuis haiti")
plt.show()

C:\Users\Dell\Miniconda3\lib\site-packages\pandas\core\frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


IndexError: index 1 is out of bounds for axis 0 with size 1

Le pic correspondant à la catastrophe est bien visible sur le graphique. Il convient d'ajouter des annotations : utiliser la commande plt.text(x,y,texte) pour ajouter une information au pic.

**Question** : Comparer le nombre d'immigrant au Canada provenant d'Inde à ceux provenant de Chine de 1983 à 2013

Conclusion : 

Comparer maintenant le flux migratoire de 1983 à 2013 des cinq pays ayant eu le plus de migrants vers le Canada sur les 20 ans.

**2. Area Plot (Graphique en aires empilées)**

Plus lisible que le dernier graphique, le graphique en aires empilées (stacked line plot ou Area plot) est préferable lorqu'on représente plusieurs séries sur la même figure.<br>
Le *Area plot* a une transparence réglée par défaut à 0,5 et e paramètre *stacked* permet de définir si les aires sont empilées ou non (True par défaut).

**Question** : Représenter avec une transparence de 0.45 les 5 pays ayant le moins migrants au Canada sur la période 1980 - 2013

**3. Les histogrammes**

Un histogramme est un moyen de représenter la distribution de fréquence d'un jeu de données numériques.<br>
Fonctionement : l’axe des x est découpé en intervalles, puis on compte le nombre de points de données attribués à chaque intervalle.<br>
L'axe des ordonnées est donc la fréquence ou le nombre de points de données dans chaque groupe.<br>
Le nombre d'intervalles **bins** (10 par défaut) est paramétrable et permet d'affiner la représentation.

Quelle est la distribution de fréquence du nombre de nouveaux immigrants de divers pays au Canada en 2013 ?

2 variables sont définies : nb et bornes_sup

Concrétement, numpy a découpé les valeurs en 10 intervalles de mêmes amplitudes (3412.9) et on compte 178 observations comprises entre 0 et 3412.9

La plupart du temps les graduations ne sont pas alignées avec le découpage des intervalles, mais il est possible d'améliorer le graphique en définissant le paramètre xticks

**Exercice**
Créer un histogramme représentant les flux migratoires vers le Canada depuis le Danemark, la norvège et la Suéde

### Diagramme à batons (Bar Charts)
La longueur de chaque baton (barre) est proportionnelle à l'effectif/la mesure associée à une caractéristique.

Deux types de diagrammes sont possibles :

- kind=bar graphique vertical
- kind=barh graphique horizontal

**1. Diagramme à batons vertical**

L'axe des x représente les caracrtéistiques. Ce type de diagramme est très utile pour représenter des séries temporelles mais son principal inconvénient est le manque de palce pour ajouter du texte en dessous de chaque baton.

Exemple : 

L'islande a connu une grave crise financière de 2008 et les effets sont clairement visibles sur le graphique. On voit, en effet, une nette augmentation du nombre d'immigrants islandais après 2008.

L'idée, ici, est d'ajouter une flèche entre la barre 2008 et celle de 2013.

En utilisant la méthode `annotate` et les paramètres suivants :
- `s`: str, ttexte de l'annotation.
- `xy`: Tuple :  coordonnées du point à annoter(pointe de la flèche).
- `xytext`: Tuple : coordonnées du point de départ du texte (départ de la flèche).

- `arrowprops`: Dictionnaire de définition de la flèche:
    - `arrowstyle`: Style de la flèche, `'->'` : pointe standard.
    - `connectionstyle`: Type de trait. `arc3` est une ligne droite.
    - `color`: Couleur de la flèche.
    - `lw`: Epaisseur du trait.

Plus d'infos sur : 
http://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.annotate.

Ajouter un texte le long de la flèche : Crise financière de 2008 - 2011.
Paramètres à utiliser : 

rotation: Rotation du texte (sens direct)
va: alignement vertical [‘center’ | ‘top’ | ‘bottom’ | ‘baseline’]
ha: alignement horisontal [‘center’ | ‘right’ | ‘left’]

**2. Diagramme à batôns horizontal**

Lorsque les étiquettes de l'axe des abscisses sont des textes nécessitant plus de place, il est plus confortable d'utiliser les diagrammes horisontaux.


**Question:** 
A partir du dataFrame `df_can` dataset, créer un diagramme à bâtons *horizontal* représentant les 15 pays ayant eu le plus grand nombre *total* d'immigrants au Canada, de 1980 à 2013. Faire apparaître le nombre d'immigrants sur les barres.
*Indication* : Faire une boucle sur les pays et utiliser la méthode annotate

<hr>
Copyright &copy; 2018 Hatem & Driss @NEEDEMAND